src/jvmMain/kotlin/io/beatmaps/login/discord.kt

```diff
@@ -20,25 +19,30 @@ import org.jetbrains.exposed.sql.and
 import org.jetbrains.exposed.sql.select
 import org.jetbrains.exposed.sql.transactions.transaction
 
-val discordProvider = OAuthServerSettings.OAuth2ServerSettings(
+fun discordProvider(state: String?) = OAuthServerSettings.OAuth2ServerSettings(
     name = "discord",
     authorizeUrl = "https://discord.com/api/oauth2/authorize",
     accessTokenUrl = "https://discord.com/api/oauth2/token",
     clientId = System.getenv("DISCORD_CLIENTID") ?: "",
     clientSecret = System.getenv("DISCORD_CLIENTSECRET") ?: "",
     requestMethod = HttpMethod.Post,
-    defaultScopes = listOf("identify")
+    defaultScopes = listOf("identify"),
+    authorizeUrlInterceptor = {
+        state?.let {
+            this.parameters["state"] = state
+        }
+    },
 )
 
-class SimpleUserPrincipal(val user: UserDao, val alertCount: Int) : Principal
+class SimpleUserPrincipal(val user: UserDao, val alertCount: Int, val redirect: String) : Principal
 
 fun Application.installDiscordOauth() {
     val baseName = System.getenv("BASE_URL") ?: Config.basename
```

Comment:
Right here ^

Reply:
Oops, ktlint deleted according to local uncommitted changes XD